test/abstract_syntax_tree_location_data.coffee

```diff
@@ -1,8 +1,29 @@
-testExpression = expressionAstMatchesObject
+# Astract Syntax Tree location data
+# ---------------------------------
+
+testAstLocationData = (code, expected) ->
+  ast = CoffeeScript.compile code, ast: yes
+  # Pull the node we’re testing out of the root `Block` node’s first child.
+  node = ast.expressions[0]
```

Comment:
So this was the most problematic thing I ran into rebasing my `Value` AST branch onto this branch: I had written location data tests for eg `a.b` that include location data for nested nodes (not just the root node)

And now I notice that per my other comment that `node = ast.expressions[0]` is not very "solid", I don't think it's good to separately hard-code it here. What exactly was gained by removing those shared test helpers?

If it's about having the nicer error messages here, you could probably refactor so that it recursively looked for nodes with location data in the "expected" location data object and then applied the individual location data field tests (with custom error messages) to each node with location data that it found?

Let me know if you'd rather do that refactoring here or if I should come up with something in the `Value` AST PR

Reply:
My concern was about code organization, of keeping all the AST stuff close together rather than adding more functions to `helpers.coffee` where those functions are only used by the two AST files. We could move them into `test/support/abstract_syntax_tree_helpers.coffee`, I suppose, and update `Cakefile` to load them like it currently makes the generic helpers global. Another option would be to just have one big AST tests file, rather than a separate one for location data tests, and then all the helpers and tests are in the same file.

Reply:
Ok I think I like option 1 (`test/support/abstract_syntax_tree_helpers.coffee`) at this point, as I anticipate both the main and location data AST test files getting quite large so seems nice in theory to keep them separate

Updated https://github.com/jashkenas/coffeescript/pull/5107 (including `test/support/abstract_syntax_tree_helpers.coffee`) to support the test cases there

Reply:
Maybe let’s do merge in the test helpers refactor (the most recent one) in the Value PR? It doesn’t really matter which one it goes in, but I think it might make more sense to do it there because that’s the PR that necessitates it.

Then I think we can merge this in? And I’ll take one last look at the PR on the CoffeeScript repo, but I assume that that’s probably ready to merge in too.

Reply:
Yup that's where I did the refactor in the Value PR (on `value-ast` branch)